datafusion/physical-optimizer/src/aggregate_statistics.rs

```diff
@@ -140,155 +138,24 @@ fn take_optimizable_column_and_table_count(
-                    if exprs.len() == 1 {
-                        // TODO optimize with exprs other than Column
-                        if let Some(col_expr) =
-                            exprs[0].as_any().downcast_ref::<expressions::Column>()
-                        {
-                            if let Precision::Exact(val) =
-                                &col_stats[col_expr.index()].min_value
-                            {
-                                if !val.is_null() {
-                                    return Some((
-                                        val.clone(),
-                                        agg_expr.name().to_string(),
-                                    ));
-                                }
-                            }
-                        }
-                    }
-                }
-            }
-            _ => {}
+    if !agg_expr.is_distinct() {
+        if let Some((val, name)) = take_optimizable_value_from_statistics(agg_expr, stats)
+        {
+            return Some((val, name));
         }
     }
     None
 }
 
 /// If this agg_expr is a max that is exactly defined in the statistics, return it.
-fn take_optimizable_max(
+fn take_optimizable_value_from_statistics(
     agg_expr: &AggregateFunctionExpr,
     stats: &Statistics,
 ) -> Option<(ScalarValue, String)> {
-    if let Precision::Exact(num_rows) = &stats.num_rows {
-        match *num_rows {
-            0 => {
-                // MIN/MAX with 0 rows is always null
-                if is_max(agg_expr) {
-                    if let Ok(max_data_type) =
-                        ScalarValue::try_from(agg_expr.field().data_type())
-                    {
-                        return Some((max_data_type, agg_expr.name().to_string()));
-                    }
-                }
-            }
-            value if value > 0 => {
-                let col_stats = &stats.column_statistics;
-                if is_max(agg_expr) {
-                    let exprs = agg_expr.expressions();
-                    if exprs.len() == 1 {
-                        // TODO optimize with exprs other than Column
-                        if let Some(col_expr) =
-                            exprs[0].as_any().downcast_ref::<expressions::Column>()
-                        {
-                            if let Precision::Exact(val) =
-                                &col_stats[col_expr.index()].max_value
-                            {
-                                if !val.is_null() {
-                                    return Some((
-                                        val.clone(),
-                                        agg_expr.name().to_string(),
-                                    ));
-                                }
-                            }
-                        }
-                    }
-                }
-            }
-            _ => {}
-        }
-    }
-    None
+    let value = agg_expr.fun().value_from_stats(
```

Comment:
We have `agg_expr.is_distinct()` here, so you can differentiate distinct count and non-distinct case.

Maybe we can have `StatisticsArgs` similar to `AccumulatorArgs`.

```rust
pub struct StatisticsArgs<'a> {
    statistics: &'a Statistics,
    return_type: &'a DataType,
    /// Whether the aggregate function is distinct.
    ///
    /// ```sql
    /// SELECT COUNT(DISTINCT column1) FROM t;
    /// ```
    pub is_distinct: bool,
    /// The physical expression of arguments the aggregate function takes.
    pub exprs: &'a [Arc<dyn PhysicalExpr>],
}
```

Reply:
So the function for count would return None if the aggregate is distinct?

Reply:
> It felt like leaking in the UDF logic required by physical optimiser

I think it makes sense, by default it has None, but we can also tune the function for optimizer 🤔

Reply:
Maybe we need to find a better name than `value_from_stats`? I am not very clear why the `distinct` for count is important if you are simply "getting the value" from stats, if the precision is Exact

It seems that distinct has specifically to do with the optimizer and not with "getting a value for the specific UDF from statistics"

Reply:
The optimization for count is mainly for `count(*)`, since it doesn't make sense for distinct case, so we only care about `non-distinct count` for `value_from_stats`, but not `distinct count`. 

> It seems that distinct has specifically to do with the optimizer and not with "getting a value for the specific UDF from statistics"

`distinct` is for `value_from_stats` to know what kind of function we have, either `count` or `distinct count`.